File: 프로그래머스/3/43238. 입국심사/입국심사.py
def solution(n, times):
    # 모든 사람이 심사 받는데 걸리는 시간의 최솟값
    # (이분 탐색): T라는 시간이 주어졌을 때, 모든 심사관이 최대 몇 명을 심사할 수 있는가?
    answer = 0
    times.sort();
    
    low, high = 0, times[-1] * n
    while low <= high:
        mid = (low + high) // 2
        total = 0
        
        for time in times:
            total += mid // time
            if total >= n: break
            
        if total >= n:
            answer = mid # n명 이상이면 정답 후보
            high = mid - 1
        else:
            low = mid + 1
            
    return answer 

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    // t시간 동안 심사관들이 총 몇 명을 심사할 수 있는지! n보다 크면서 가장 작은 t 구하기 -> t를 정하는 이분 탐색 진행
    sort(times.begin(), times.end());
    
    long long low = 1;
    long long high = (long long)times.back() * n;
    
    while (low <= high) {
        long long mid = (low + high) / 2;
        long long total = 0;
        
        for (auto t : times) {
            total += (mid / t);
            if (total >= n) break;
        }
        
        if (total >= n) {
            answer = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    
    return answer;
}